### PR TITLE
Keep all laser scan planes parallel to each other by changing the Y value, instead of the pitch of the planes

### PR DESCRIPTION
**Pull request by tckarenchiang**
This is related to #10. 
Now the laser scanners are mounted at different origins, so the beams won't cross each other

### CODE REVIEW

File: modROS.lua [hunk 1]
```diff
@@ -336,8 +332,8 @@ function ModROS:publish_laser_scan_func()
         -- serialise to JSON and write to pipe
         self.file_pipe:write(sensor_msgs_LaserScan.ros_msg_name .. "\n" .. scan_msg:to_json())
 
-        -- calculate the angle of the current layer's scanning plane wrt origin of the laser
-        local scanning_plane_pitch = -math.atan(delta_y * i / radius)
+        -- the angle of the current layer's scanning plane wrt origin of the laser
+        local scanning_plane_pitch = 0
         -- convert to quaternion for ROS TF; scanning_plane_pitch is the angle rotated from the origin of the laser scanner along y axis.
         local q = ros_quaternion.from_euler(0, scanning_plane_pitch, 0)
```
Comment (gavanderhoorn) [inline modROS.lua:338]: Seeing as there is no inclination any more of the planes, I'd suggest removing this constant.

```suggestion
        -- convert to quaternion for ROS TF
        local q = ros_quaternion.from_euler(0, 0, 0)
```

something to consider: do we even need to convert anything here?

Could just set a `[0, 0, 0, 1]` quaternion later in `tf_base_link_laser_frame_i`.

File: modROS.lua [hunk 2]
```diff
@@ -282,32 +282,28 @@ end
...
-    -- And "delta_y" is added between the scanning planes, along +y direction from the origin of the laser scanner
-    local delta_y_total = 0.05
-    -- quarter of length as there are four segements between 5 laser scan layers
-    local delta_y = delta_y_total / (mod_config.laser_scan.num_layers - 1)
     for i = 0, mod_config.laser_scan.num_layers-1 do
         self.laser_scan_array = {}
+        -- get the cooridante(world) of each laser scanner's origin
+        -- "mod_config.laser_scan.inter_layer_distance" is added between the scanning planes, along +y direction from the lowest laser scan plane
+        -- and all laser scan planes are parallel to each other
+        local orig_x, orig_y, orig_z = localToWorld(self.instance_veh.cameraNode, 0, mod_config.laser_scan.inter_layer_distance * i, 0)
```
Comment (gavanderhoorn) [inline modROS.lua:297]:
```suggestion
        local laser_dy = mod_config.laser_scan.inter_layer_distance * i
        local orig_x, orig_y, orig_z = localToWorld(self.instance_veh.cameraNode, 0, laser_dy, 0)
```

could then use `laser_dy` later on when setting up `tf_base_link_laser_frame_i` as well.

(name could be different. Perhaps something else would make more sense)

File: modROS.lua (hunk 1 above)
Comment (tckarenchiang) [reply modROS.lua:338]: we don't actually. I thought it would be nice to keep it but agreed, it's better to just remove it.

File: modROS.lua [hunk 3]
```diff
@@ -282,32 +282,28 @@ end
...
     local delta_theta = LS_FOV / (mod_config.laser_scan.num_rays - 1)
-    -- the "delta_y_total = 0.05" varaible is the total distance from the z axis of non-tilted laser scan layer to the z axis of last laser scan layer (most tilted one)
-    -- and is to determine how much the laser scan layers are tilted, the larger the length is, the more titlted the laser scan layer is
-    -- And "delta_y" is added between the scanning planes, along +y direction from the origin of the laser scanner
-    local delta_y_total = 0.05
-    -- quarter of length as there are four segements between 5 laser scan layers
-    local delta_y = delta_y_total / (mod_config.laser_scan.num_layers - 1)
     for i = 0, mod_config.laser_scan.num_layers-1 do
         self.laser_scan_array = {}
+        -- get the cooridante(world) of each laser scanner's origin
```
Comment (gavanderhoorn) [inline modROS.lua:294]:
```suggestion
        -- get the (world) coordinate of each laser scanner's origin
```

File: modROS.lua [hunk 4]
```diff
@@ -354,7 +349,7 @@ function ModROS:publish_laser_scan_func()
             -- note the order of the axes here (see earlier comment about FS chirality)
             base_to_laser_z,
             base_to_laser_x,
-            base_to_laser_y,
+            base_to_laser_y + laser_dy * i,
```
Comment (gavanderhoorn) [inline modROS.lua:352]: Doesn't `laser_dy` already contain the total offset?
```suggestion
            base_to_laser_y + laser_dy,
```

Comment (tckarenchiang) [reply modROS.lua:352]: oops yes!

**Review by gavanderhoorn — APPROVED** [review]
I haven't ran it in FS19, but I assume you have @tckarenchiang.

I'll leave merging to you. Perhaps we'd like to wait for @Sagexs to confirm it works for him as well?

**Comment by tckarenchiang** [comment]
> I haven't ran it in FS19, but I assume you have @tckarenchiang.
> 
Yes, I've ran it in farmsim.

**Comment by Sagexs** [comment]
sry for the late repspond, was on holiday. Things looks fine to me. Great work you two!! only thing thats missing are the crops and vegetation in the scans. 

Do you see any possibility of displaying them in the laser scans? I guess that they do not have a collision box and are therefore not currently displayed. 

fs192
correct

**Comment by gavanderhoorn** [comment]
@Sagexs: could you please post your comment on #10 instead of this PR?

That way we keep the discussion in one place.

**Pull request merged**